### ISSUE TITLE
clamav anti-virus is memory intensive

### ISSUE THREAD

**Issue by eeelllccc** [alan-turing-institute/data-safe-haven#1724]
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a problem observed when **using** a Data Safe Haven.
- [ ] I can reproduce this with the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :computer: System information

- **Data Safe Haven version:** 4.1.0 
- **Browser details:** N/A 
- **Operating system details:** 
"type": "Ubuntu",
        "version": "20.04.2023082900" 

## :no_entry_sign: Describe the problem

During the Dec 2023 DSG we found that participants were having out of memory issues with some VMs. The clamav anti-virus software was using a significant amount of memory (around 25%) on the standard D2 v3 VMs with 8GB memory.

@JimMadge commented that 8GB memory is probably not enough for several users (even for light usage), which we have taken on board. 

Raising this as an issue to check that this memory usage by clamav is as expected.

## :steam_locomotive: Workarounds or solutions

Use machines with more memory!

**Comment by JimMadge** [comment]
Would be helpful to figure out which process is causing the problem.

I think there should be two. The`clamonacc` on-access daemon, and the `clamscan` periodic, full-system scan.

**Comment by craddm** [comment]
It must be related to `clamdscan`, as `clamonacc` would not have been running, as per #1722

**Comment by JimMadge** [comment]
@craddm Can you reproduce this?

If that is true (not 100% certain but that seems likely), `clamscan` is set to run daily at 01:00 ([timer](https://github.com/alan-turing-institute/data-safe-haven/blob/5600650e987c00ddece5d0660a2e3c9d013eb869/deployment/common/resources/clamav-clamdscan.timer)).

Is it running for a very long time?
Might want to tighten up the frequency (is daily a DSPT requirement?) and the scan locations ([currently `/`](https://github.com/alan-turing-institute/data-safe-haven/blob/5600650e987c00ddece5d0660a2e3c9d013eb869/deployment/common/resources/clamav-clamdscan.service)).

**Comment by craddm** [comment]
During our chat with TRESA yesterday, I seem to remember seeing the actual error reproduced at some point. Could somebody provide a screenshot? @helendduncan @cptanalatriste @dsj976

**Comment by helendduncan** [comment]
<img width="1726" alt="Screenshot 2024-02-01 at 12 32 44" src="https://github.com/alan-turing-institute/data-safe-haven/assets/46891265/9529d8c3-00e6-4d8e-abd7-f058450c0132">

Is this what you meant @craddm

**Comment by helen-d-little** [comment]
I think we saw the proxy error, I don't remember the memory one @craddm ?

**Comment by craddm** [comment]
> <img alt="Screenshot 2024-02-01 at 12 32 44" width="1726" src="https://private-user-images.githubusercontent.com/46891265/301504849-9529d8c3-00e6-4d8e-abd7-f058450c0132.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY3OTExNTgsIm5iZiI6MTcwNjc5MDg1OCwicGF0aCI6Ii80Njg5MTI2NS8zMDE1MDQ4NDktOTUyOWQ4YzMtMDBlNi00ZDhlLWFiZDctZjA1ODQ1MGMwMTMyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjAxVDEyMzQxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFiM2M1OWZhN2E5OGZkODEzMmUxNGQ2YjQzMWNiNDNmYTA3MjAzMjAyNDBjMjA1YmQ3YTVmZGRlNGM3MWE4YjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.7bbGyvi8jINEOB9FhnU0Jp4l0jc9WH-6R44NFEtGR2g">
> Is this what you meant @craddm

No, that's the issue with the squid proxy. I seem to remember there being *another* one that up briefly.

> I think we saw the proxy error, I don't remember the memory one @craddm ?

I might be misremembering, but I thought at some point someone was logged in to an SRD and seeing some error that mentioned clamav, before moving on to the proxy issue. But again, I might be misremembering. Fundamentally I need to see somewhere the problem is happening, as it doesn't seem to be happening on any SRDs I have up.

**Comment by craddm** [comment]
> Is it running for a very long time? Might want to tighten up the frequency (is daily a DSPT requirement?) and the scan locations ([currently `/`](https://github.com/alan-turing-institute/data-safe-haven/blob/5600650e987c00ddece5d0660a2e3c9d013eb869/deployment/common/resources/clamav-clamdscan.service)).

The latest DSPT guidance says "Antivirus/anti-malware is kept continually up to date.", but doesn't specify a specific time period beyond _continually_

**Comment by craddm** [comment]
Again, I might have totally misremembered any sort of `clamav` error showing up yesterday, so maybe you can disregard that.

So, I've seen some suggestions that 1-2 Gb for `clamav` is not that unusual:

https://github.com/Cisco-Talos/clamav/issues/565

Hah, actually scratch that about not seeing it that high on any of my currently running SRDs - 

![image](https://github.com/alan-turing-institute/data-safe-haven/assets/5796417/531a2892-8120-4dc8-91e2-5505bdfe27bd)

**Comment by eeelllccc** [comment]
That looks familiar, I think it was a `clamav-daemon` process that I originally spotted causing our problems.

**Comment by jemrobinson** [comment]
@craddm : Is this fixed in 4.2.0?

**Comment by craddm** [comment]
No. I've had another look with the SRDs from the current release, and can see memory usage consistent with the report above - approx 2 Gb. 

I think the memory usage of `clamav-daemon`/`clamd` is not avoidable - it keeps the virus signature database in memory at all times, and so the memory usage just reflects the size of its database. The on-access scanner requires `clamd` to be running to work, and on-access scanning is a DSPT requirement

This is separate from the daily `clamscan` run, which doesn't require `clamd` to be running.

Basically, I don't think there's anything we can do about it.

**Comment by JimMadge** [comment]
If it is a problem or not necessary, we could stop the daemon.

Otherwise, maybe we should add some advice about suitable VM sizes noting that the memory overhead of essential processes will be ~2-3GB.

**Comment by JimMadge** [comment]
Just found that clamd provides `clamdtop` which describes clamd process including memory usage :tada:.

**Comment by JimMadge** [comment]
On v5.0.x,

```console
  ClamdTOP version 0.103.12   Mon Sep 30 13:45:57 2024
NO CONNTIME LIV IDL QUEUE  MAXQ   MEM ENGINE  DBVER DBTIME        HOST
 1 00:02:52   1   0     0     0 1.38G 0.103.12 27410 2024-09-26T09 local
```

1.38 GiB memory usage idle.

Also noting [this section](https://docs.clamav.net/manual/Installing/Docker.html?highlight=memory#memory-ram-requirements) of the (Docker 😕) documentation.

> Recommended RAM for ClamAV (As of 2020/09/20):
> 
> Minimum: 3 GiB
> Preferred: 4 GiB


> ClamAV uses upwards of 1.2 GiB of RAM simply to load the signature definitions into matching structures in the construct we call an "engine". This does not take into account any RAM required to process the files during the scanning process.
> 
> ClamAV uses upwards of 2.4 GiB of RAM for a short period each day when loading new signature definitions. When the clamd processs reloads the databases after an update, the default behavior is for ClamAV to build a new engine based on the updated signatures first. Once loaded and once all scans that use the old engine have completed, the old engine is unloaded. This process is called "concurrent reloading" and enables scans to continue during the reload. As a consequence, clamd will use twice the amount of RAM for a brief period. During the reload.
> 
> The freshclam process may also consume a sizeable chunk of memory when load-testing newly downloaded databases. It won't use quite as much as a clamd database reload, but it may still be enough to cause issues on some systems.

So, in summary, the memory usage we can see now is about what we would expect. It will double for some length of time daily while the signatures are updated, as both old and new signatures are held until the old signatures are no longer needed.

We could change that behaviour with

```
ConcurrentDatabaseReload no
```

in `clamd.conf`. But that would mean no scans happen while the signature database is updated. My reading of the docs is they will be held until after the update has finished.

That would make on access scanning less responsive (or not work at all?), which is a DSPT requirement.

I think we should not make any code changes here.
We could add a note to the documentation about recommended VM sizes for workspaces that notes this though.

Any thoughts/disagreements @craddm @jemrobinson ?

**Comment by jemrobinson** [comment]
Are we at a point where our constraints (needs to support premium storage, needs at least X GB of RAM) gives a relatively short list of suggested VM sizes? If so, should we add this suggested list to the docs?

**Comment by craddm** [comment]
I think that's reasonable @JimMadge - I don't think there's anything unexpected or avoidable about the memory use.

**Comment by JimMadge** [comment]
I haven't found a table where I can filter on the things we are interested in from all VM sizes. I think that list would end up being quite long.

I think there are other factors too. D series feels like a reasonable start, but you might want to go memory-focused or CPU-focused or even HPC if you need it.
I'll try writing some helpful suggestions.
